src/option.ts

```diff
@@ -1,21 +1,168 @@
+import { Any, MapFn } from "./types";
+import { AsyncResult, Err, Ok, promiseOfResultToAsyncResult, Result } from "./result";
+
+export type FlatMapFn<T, T2> = (value: T) => Option<T2> | Promise<Option<T2>>;
+
+export type Option<T> = Some<T> | None;
+
+export type AsyncOption<T> = {
+    value: Promise<Option<T>>;
+    map<R>(f: MapFn<T, R>): R extends Promise<infer R2> ? AsyncOption<R2> : AsyncOption<R>;
+} & Omit<Option<T>, "value" | "map" | "mapError" | "flatMap"> &
+    Promise<Option<T>>;
+
 export type Some<T> = {
     some: true;
     value: T;
+    map<R>(f: MapFn<T, R>): R extends Promise<infer R2> ? AsyncSome<R2> : Some<R>;
+    flatMap<T2>(
+        f: FlatMapFn<T, T2>,
+    ): ReturnType<typeof f> extends Promise<T2> ? AsyncOption<T2> : Option<T2>;
+    attemptMap<R>(
+        f: MapFn<T, R>,
+    ): R extends Promise<infer R2> ? AsyncResult<R2, unknown> : Result<R, unknown>;
 };
 
 export type None = {
     some: false;
     value: never;
+    map: <R>(f: MapFn<never, R>) => None;
+    flatMap: <T2>(f: FlatMapFn<never, T2>) => None;
+    attemptMap: <T2>(f: MapFn<never, T2>) => Result<None, never>;
```

Comment:
In Some attemptMap returns a Result<R, unknown> and in None it returns Result<None, never>.
R is the value and None is a monad. Maybe the return type should just be Ok<undefined>, which reflects what will actually happen. The function will never get run, so it will neither fail nor return anything.